### PR TITLE
Autoinstrumentation for Struts2 actions

### PR DESCRIPTION
**Pull request by vovencij**
This PR adds autoinstrumentation for [Struts2 MVC framework](https://struts.apache.org/index.html). It does 2 things:

1. Create an internal span for Struts action invocations while catching exceptions properly (by default struts will catch all exceptions and report that a http call was successful, even if action handling exploded). Internal span has properly filled `code.namespace` and `code.function`.
2. For configuration with parameter values encoded in servlet path, update server span name with properly encoded variable names as specified in struts configuration.

Instrumentation doesn't support Struts 1, which is EOL-ed long ago.

Screenshot 2020-11-13 at 11 57 23

### CODE REVIEW

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      // normalize parameters, not actions.
+      return;
+    }
+
+    getCurrentServerSpan().updateName(ServletContextPath.prepend(context, result));
+  }
+
+  @Override
+  protected String getInstrumentationName() {
+    return "io.opentelemetry.auto.struts-2";
```
Comment (trask) [inline instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java:74]:
```suggestion
    return "io.opentelemetry.auto.struts-2.0";
```

File: settings.gradle [hunk 2]
```diff
@@ -188,6 +188,7 @@ include ':instrumentation:spring:starters:otlp-exporter-starter'
 include ':instrumentation:spring:starters:zipkin-exporter-starter'
 include ':instrumentation:spymemcached-2.12'
 include ':instrumentation:twilio-6.6'
+include ':instrumentation:struts-2'
```
Comment (trask) [inline settings.gradle:191]: also, can you send a separate PR adding struts to the main README? (separate PR so we can wait to update the README when we release the next version)

```suggestion
include ':instrumentation:struts-2.0'
```

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    Class<?> actionClass = action.getClass();
+
+    String method = actionInvocation.getProxy().getMethod();
+    String spanName = spanNameForMethod(actionClass, method);
+
+    Span strutsSpan = tracer.spanBuilder(spanName).startSpan();
+
+    strutsSpan.setAttribute("code.namespace", actionClass.getName());
+    if (method != null) {
+      strutsSpan.setAttribute("code.function", method);
```
Comment (trask) [inline instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java:30]: can you submit a PR to opentelemetry-java to add these to `SemanticAttributes`?

File: instrumentation/struts-2/struts-2.gradle [hunk 4]
```diff
@@ -0,0 +1,27 @@
+apply from: "$rootDir/gradle/instrumentation.gradle"
+
+muzzle {
+  pass {
+    group = "org.apache.struts"
+    module = "struts2-core"
+    versions = "[2.3.20,)"
```
Comment (trask) [inline instrumentation/struts-2/struts-2.gradle:7]: is this the minimum version that this instrumentation works with? if so, then let's rename module to `struts-2.3` instead of my suggestion below of `struts-2.0`

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2InstrumentationModule.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.util.List;
+
+@AutoService(InstrumentationModule.class)
+public class Struts2InstrumentationModule extends InstrumentationModule {
+
+  public Struts2InstrumentationModule() {
+    super("struts-2");
```
Comment (trask) [inline instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2InstrumentationModule.java:19]: based on recent discussion in #1579
```suggestion
    super("struts", "struts-2.0");
```

**Review by trask — COMMENTED** [review]
thanks @vovencij!

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import static io.opentelemetry.javaagent.instrumentation.struts2.Struts2Tracer.TRACER;
+
+import com.opensymphony.xwork2.ActionInvocation;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge;
+import net.bytebuddy.asm.Advice;
+
+public class ActionInvocationAdvice {
```
Comment (anuraaga) [inline instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java:16]: It looks like struts has the concept of interceptors

https://struts.apache.org/core-developers/interceptors.html

Is it possible to define an interceptor and register it, instead of instrumenting the action invocation directly? While it wouldn't need to be in this PR, it would make it much easier to separate out library, nonagent instrumentation.

File: instrumentation/struts-2/struts-2.gradle (hunk 4 above)
Comment (vovencij) [reply instrumentation/struts-2/struts-2.gradle:7]: 2.3.20 from year 2014 is the earliest version of Struts2, that is available for download on Struts official download page: https://archive.apache.org/dist/struts/ I wasn't adventurous enough to search for even earlier versions. :)

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java (hunk 1 above)
Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: Renamed struts-2 to struts-2.3 everywhere.

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2InstrumentationModule.java (hunk 5 above)
Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2InstrumentationModule.java]: Renamed struts-2 to struts-2.3 everywhere.

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java (hunk 6 above)
Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java]: Yes, interceptors definitely can be used for a library. Moved current instrumentation to `javaagent` sub-project to accommodate for the library version. Which I will do as a separate PR.

File: settings.gradle (hunk 2 above)
Comment (vovencij) [reply settings.gradle]: Updated version to 2.3. Main README PR: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1636

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java (hunk 3 above)
Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: Will do.

Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: Do I understand correctly, that to update `SemanticAttributes.java` I have to update https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic_conventions/trace/general.yaml from which `SemanticAttributes.java` s eventually generated?

Comment (trask) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: oh yes, good find, I believe that's the correct spot

**Review by anuraaga — APPROVED** [review]
Thanks!

File: instrumentation/struts-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static void onEnter(
+      @Advice.This ActionInvocation actionInvocation,
+      @Advice.Local("otelSpan") Span span,
+      @Advice.Local("otelScope") Scope scope) {
+    span = TRACER.startSpan(actionInvocation);
+    scope = TRACER.startScope(span);
+
+    TRACER.updateServerSpanName(Java8BytecodeBridge.currentContext(), actionInvocation.getProxy());
```
Comment (iNikem) [inline instrumentation/struts-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/ActionInvocationAdvice.java:26]: We [have recently switched](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1540) to using `tracer()`

File: instrumentation/struts-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    if (!result.contains("{")) {
+      // If there are no braces, then there are no path parameters encoded in
+      // the action name, so let's not change existing server span name, because
+      // path is good enough. Wildcards like * in action name may glue
+      // several endpoints into one action name, which we do not want -- we want
+      // normalize parameters, not actions.
+      return;
+    }
+
+    getCurrentServerSpan().updateName(ServletContextPath.prepend(context, result));
```
Comment (iNikem) [inline instrumentation/struts-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java:69]: You already have this as `serverSpan` local variable.

File: instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    trace.span(index) {
+      name expectedHandlerName(endpoint)
+      kind Span.Kind.INTERNAL
+      errored endpoint == EXCEPTION
+      if (endpoint == EXCEPTION) {
+        errorEvent(Exception, EXCEPTION.body)
+      }
+      attributes {
+        'code.namespace' "io.opentelemetry.struts.GreetingAction"
+        'code.function' { it == expectedMethodName(endpoint) }
```
Comment (iNikem) [inline instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy:81]:
```suggestion
        'code.function' expectedMethodName(endpoint)
```

File: instrumentation/struts-2.3/javaagent/struts-2.3-javaagent.gradle [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    group = "org.apache.struts"
+    module = "struts2-core"
+    versions = "[2.3.20,)"
+  }
+}
+
+dependencies {
+  compileOnly group: 'org.apache.struts', name: 'struts2-core', version: '2.3.20'
+
+  testImplementation group: 'org.apache.struts', name: 'struts2-core', version: '2.3.20'
```
Comment (iNikem) [inline instrumentation/struts-2.3/javaagent/struts-2.3-javaagent.gradle:14]:
```suggestion
  library group: 'org.apache.struts', name: 'struts2-core', version: '2.3.20'
```

File: settings.gradle [hunk 11]
```diff
@@ -188,6 +188,7 @@ include ':instrumentation:spring:starters:otlp-exporter-starter'
 include ':instrumentation:spring:starters:zipkin-exporter-starter'
 include ':instrumentation:spymemcached-2.12'
 include ':instrumentation:twilio-6.6'
+include ':instrumentation:struts-2.3:javaagent'
```
Comment (iNikem) [inline settings.gradle:191]:
```suggestion
include ':instrumentation:struts-2.3:javaagent'
include ':instrumentation:twilio-6.6'
```

File: testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/AttributesAssert.groovy [hunk 12]
```diff
@@ -41,7 +41,7 @@ class AttributesAssert {
     } else if (value instanceof Closure) {
       assert ((Closure) value).call(val)
     } else {
-      assert val == value
+      assert value == val
```
Comment (iNikem) [inline testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/AttributesAssert.groovy:44]: Why?

File: instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy (hunk 9 above)
Comment (vovencij) [reply instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy:81]: Calling methods in assertions doesn't work with Groovy Magic in `io.opentelemetry.instrumentation.test.asserts.AttributesAssert#attribute`. I failed to get it to work over several attempts, as you can see from the commit history.

File: testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/AttributesAssert.groovy (hunk 12 above)
Comment (vovencij) [reply testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/AttributesAssert.groovy:44]: No particular reason. I tried to get nice assertion messages out of here, which would indicate for which attribute the assertion failed, but I failed eventually as various places across the project were failing with all the solutions I tried. So eventually I returned it in the initial implementation, but sides got switched incidentally. Will put it back as it was.

File: instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy (hunk 9 above)
Comment (vovencij) [reply instrumentation/struts-2.3/javaagent/src/test/groovy/Struts2ActionSpanTest.groovy]: But I think I found a nicer way instead of this closure.

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java (hunk 3 above)
Comment (vovencij) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: https://github.com/open-telemetry/opentelemetry-specification/pull/1229

**Review by iNikem — APPROVED** [review]

File: instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java (hunk 3 above)
Comment (iNikem) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: @trask are you Ok with merging this PR before we have spec updated, right?

Comment (trask) [reply instrumentation/struts-2/src/main/java/io/opentelemetry/javaagent/instrumentation/struts2/Struts2Tracer.java]: yes 👍
